File: core/dcp_patcher.py
"""
core/dcp_patcher.py — Dreamcast DCP patch applier
M5 Dev | GPL v3

DCP format (Universal Dreamcast Patcher compatible):
  A .dcp file is a ZIP archive containing:
    - Modified/new game files in their original disc filesystem path
    - xdelta/  (optional) — binary diff patches as <filename>.xdelta3
    - bootsector/IP.BIN  (optional) — replacement boot sector
"""

import os
import zipfile
import shutil
import subprocess
import tempfile
from typing import Callable

from core.config import get_tool_path
from core.logger import log as global_log


class DcpPatcher:
    def __init__(self, on_log: Callable = None, on_progress: Callable = None):
        self.on_log = on_log or global_log
        self.on_progress = on_progress or (lambda pct: None)

    def apply(
        self,
        dcp_path: str,
        disc_dir: str,       # directory containing extracted GDI/CUE files
        output_dir: str,
        ignore_checksum: bool = False,
    ) -> dict:
        """
        Apply a DCP patch to an extracted Dreamcast disc directory.

        Args:
            dcp_path:   Path to the .dcp patch file
            disc_dir:   Path to folder containing extracted disc files
                        (track03.bin and game filesystem files)
            output_dir: Where to write the patched disc files
            ignore_checksum: Skip xdelta source checksum verification

        Returns:
            dict with keys: success (bool), files_patched (int),
                            xdelta_applied (int), ipbin_replaced (bool),
                            error (str or None)
        """
        if not os.path.isfile(dcp_path):
            return {'success': False, 'error': f'DCP file not found: {dcp_path}'}
        if not os.path.isdir(disc_dir):
            return {'success': False, 'error': f'Disc directory not found: {disc_dir}'}

        os.makedirs(output_dir, exist_ok=True)

        # Copy source disc to output first (non-destructive)
        self._log('[DCP] Copying source disc to output directory...')
        self._copy_disc(disc_dir, output_dir)
        self.on_progress(10.0)

        result = {
            'success': False,
            'files_patched': 0,
            'xdelta_applied': 0,
            'ipbin_replaced': False,
            'error': None,
        }

        try:
            if not zipfile.is_zipfile(dcp_path):
                result['error'] = 'DCP file is not a valid ZIP/DCP archive.'
                return result

            with zipfile.ZipFile(dcp_path, 'r') as dcp:
                names = dcp.namelist()
                total = len(names)
                self._log(f'[DCP] Archive contains {total} entries.')

                for i, name in enumerate(names):
                    pct = 10.0 + (i / total) * 80.0
                    self.on_progress(pct)

                    dest_path = os.path.abspath(os.path.join(output_dir, name))
                    base_path = os.path.abspath(output_dir)
                    if not dest_path.startswith(base_path + os.sep) and dest_path != base_path:
                        raise ValueError(f'Malicious archive entry detected (Path Traversal): {name}')

                    # bootsector/IP.BIN → replace IP.BIN in output
                    if name.lower() == 'bootsector/ip.bin':
                        ipbin_data = dcp.read(name)
                        ipbin_out = os.path.join(output_dir, 'IP.BIN')
                        with open(ipbin_out, 'wb') as f:
                            f.write(ipbin_data)
                        result['ipbin_replaced'] = True
                        self._log('[DCP] Replaced IP.BIN from patch.')
                        continue

                    # Skip xdelta folder itself (handled below)
                    if name.startswith('xdelta/') or name.startswith('xdelta\\'):
                        continue

                    # Skip directory entries
                    if name.endswith('/'):
                        continue

                    # Regular file replacement
                    target = dest_path
                    os.makedirs(os.path.dirname(target), exist_ok=True)
                    with dcp.open(name) as src, open(target, 'wb') as dst:
                        shutil.copyfileobj(src, dst)
                    result['files_patched'] += 1
                    self._log(f'[DCP] Replaced: {name}')

                # Apply xdelta patches
                xdelta_entries = [n for n in names
                                  if n.startswith('xdelta/') and not n.endswith('/')]
                if xdelta_entries:
                    xdelta_result = self._apply_xdelta_patches(
                        dcp, xdelta_entries, output_dir, ignore_checksum
                    )
                    result['xdelta_applied'] = xdelta_result
                    result['files_patched'] += xdelta_result

            self.on_progress(100.0)
            result['success'] = True
            self._log(f'[DCP] Done. Files patched: {result["files_patched"]}, '
                      f'xdelta: {result["xdelta_applied"]}, '
                      f'IP.BIN replaced: {result["ipbin_replaced"]}')
            return result

        except Exception as e:
            result['error'] = str(e)
            self._log(f'[DCP ERROR] {e}')
            return result

    def _apply_xdelta_patches(
        self,
        dcp: zipfile.ZipFile,
        xdelta_entries: list,
        output_dir: str,
        ignore_checksum: bool,
    ) -> int:
        """Apply all xdelta patches from the xdelta/ subfolder."""
        xdelta3 = get_tool_path('xdelta3')
        applied = 0

        with tempfile.TemporaryDirectory() as tmp:
            for entry in xdelta_entries:
                # xdelta/FILENAME.xdelta3 → target is FILENAME in output_dir
                basename = os.path.basename(entry)
                # Strip .xdelta3 / .xdelta / .vcdiff extension
                for ext in ('.xdelta3', '.xdelta', '.xd', '.vcdiff'):
                    if basename.lower().endswith(ext):
                        target_name = basename[: -len(ext)]
                        break
                else:
                    target_name = basename

                dest_path = os.path.abspath(os.path.join(output_dir, target_name))
                base_path = os.path.abspath(output_dir)
                if not dest_path.startswith(base_path + os.sep) and dest_path != base_path:
                    raise ValueError(f'Malicious archive entry detected (Path Traversal): {entry}')

                patch_tmp = os.path.join(tmp, basename)
                source_path = dest_path
                output_path = os.path.join(output_dir, target_name + '.patched')

                if not os.path.isfile(source_path):
                    self._log(f'[DCP WARN] xdelta source not found: {target_name} — skipping')
                    continue

                # Extract patch to temp
                with dcp.open(entry) as src, open(patch_tmp, 'wb') as dst:
                    shutil.copyfileobj(src, dst)

                args = [xdelta3, '-d']
                if ignore_checksum:
                    args.append('-n')
                args += ['-s', source_path, patch_tmp, output_path]

                try:
                    proc = subprocess.run(args, capture_output=True, text=True)
                    if proc.returncode == 0:
                        os.replace(output_path, source_path)
                        applied += 1
                        self._log(f'[DCP xdelta] Patched: {target_name}')
                    else:
                        err = proc.stderr.strip()
                        self._log(f'[DCP xdelta ERROR] {target_name}: {err}')
                except FileNotFoundError:
                    self._log('[DCP ERROR] xdelta3 binary not found.')
                    break

        return applied

    def _copy_disc(self, src_dir: str, dst_dir: str):
        """Copy all files from src_dir to dst_dir, preserving structure."""
        for root, dirs, files in os.walk(src_dir):
            rel = os.path.relpath(root, src_dir)
            dst_root = os.path.join(dst_dir, rel) if rel != '.' else dst_dir
            os.makedirs(dst_root, exist_ok=True)
            for f in files:
                shutil.copy2(os.path.join(root, f), os.path.join(dst_root, f))

    def _log(self, msg: str):
        global_log(msg)
        if self.on_log:
            try:
                self.on_log(msg)
            except Exception:
                pass

File: core/lib/src/dcp_patcher.dart
// OpenROM — Universal ROM Compression Suite
// M5 Dev | GPL v3

import 'dart:io';
import 'package:archive/archive_io.dart';
import 'package:path/path.dart' as p;
import 'config.dart' as config;
import 'logger.dart' as logger;

class DcpPatcher {
  final void Function(String msg)? onLog;
  final void Function(double pct)? onProgress;

  DcpPatcher({this.onLog, this.onProgress});

  void _log(String msg) {
    logger.log(msg);
    if (onLog != null) {
      try {
        onLog!(msg);
      } catch (_) {}
    }
  }

  void _updateProgress(double pct) {
    if (onProgress != null) {
      try {
        onProgress!(pct);
      } catch (_) {}
    }
  }

  Map<String, dynamic> apply({
    required String dcpPath,
    required String discDir,
    required String outputDir,
    bool ignoreChecksum = false,
  }) {
    if (!File(dcpPath).existsSync()) {
      return {'success': false, 'error': 'DCP file not found: $dcpPath'};
    }
    if (!Directory(discDir).existsSync()) {
      return {'success': false, 'error': 'Disc directory not found: $discDir'};
    }

    Directory(outputDir).createSync(recursive: true);

    _log('[DCP] Copying source disc to output directory...');
    _copyDisc(discDir, outputDir);
    _updateProgress(10.0);

    final result = <String, dynamic>{
      'success': false,
      'files_patched': 0,
      'xdelta_applied': 0,
      'ipbin_replaced': false,
      'error': null,
    };

    try {
      final bytes = File(dcpPath).readAsBytesSync();
      final archive = ZipDecoder().decodeBytes(bytes);

      final total = archive.length == 0 ? 1 : archive.length;
      _log('[DCP] Archive contains $total entries.');

      final basePath = p.normalize(p.canonicalize(outputDir));
      final xdeltaEntries = <ArchiveFile>[];

      for (int i = 0; i < archive.length; i++) {
        final pct = 10.0 + (i / total) * 80.0;
        _updateProgress(pct);

        final file = archive[i];
        final name = file.name;
        final destPath = p.normalize(p.join(basePath, name));

        if (!destPath.startsWith(basePath + p.separator) && destPath != basePath) {
          throw FormatException('Malicious archive entry detected (Path Traversal): $name');
        }

        if (name.toLowerCase() == 'bootsector/ip.bin') {
          final ipbinData = file.content as List<int>;
          File(p.join(outputDir, 'IP.BIN')).writeAsBytesSync(ipbinData);
          result['ipbin_replaced'] = true;
          _log('[DCP] Replaced IP.BIN from patch.');
          continue;
        }

        if (name.startsWith('xdelta/') || name.startsWith('xdelta\\')) {
          if (file.isFile) {
            xdeltaEntries.add(file);
          }
          continue;
        }

        if (!file.isFile) continue;

        File(destPath)
          ..createSync(recursive: true)
          ..writeAsBytesSync(file.content as List<int>);
        result['files_patched'] = (result['files_patched'] as int) + 1;
        _log('[DCP] Replaced: $name');
      }

      if (xdeltaEntries.isNotEmpty) {
        final xdeltaApplied = _applyXdeltaPatches(
          xdeltaEntries,
          outputDir,
          ignoreChecksum,
        );
        result['xdelta_applied'] = xdeltaApplied;
        result['files_patched'] = (result['files_patched'] as int) + xdeltaApplied;
      }

      _updateProgress(100.0);
      result['success'] = true;
      _log('[DCP] Done. Files patched: ${result["files_patched"]}, '
          'xdelta: ${result["xdelta_applied"]}, '
          'IP.BIN replaced: ${result["ipbin_replaced"]}');
      return result;
    } catch (e) {
      result['error'] = e.toString();
      _log('[DCP ERROR] $e');
      return result;
    }
  }

  int _applyXdeltaPatches(
    List<ArchiveFile> xdeltaEntries,
    String outputDir,
    bool ignoreChecksum,
  ) {
    final xdelta3 = config.getToolPath('xdelta3');
    int applied = 0;

    final tempDir = Directory.systemTemp.createTempSync('openrom_dcp_');
    try {
      final basePath = p.normalize(p.canonicalize(outputDir));

      for (final entry in xdeltaEntries) {
        final basename = p.basename(entry.name);
        String targetName = basename;

        for (final ext in ['.xdelta3', '.xdelta', '.xd', '.vcdiff']) {
          if (basename.toLowerCase().endsWith(ext)) {
            targetName = basename.substring(0, basename.length - ext.length);
            break;
          }
        }

        final destPath = p.normalize(p.join(basePath, targetName));
        if (!destPath.startsWith(basePath + p.separator) && destPath != basePath) {
          throw FormatException('Malicious archive entry detected (Path Traversal): ${entry.name}');
        }

        final patchTmp = p.join(tempDir.path, basename);
        final sourcePath = destPath;
        final outputPath = '$destPath.patched';

        if (!File(sourcePath).existsSync()) {
          _log('[DCP WARN] xdelta source not found: $targetName — skipping');
          continue;
        }

        File(patchTmp).writeAsBytesSync(entry.content as List<int>);

        final args = [xdelta3, '-d'];
        if (ignoreChecksum) {
          args.add('-n');
        }
        args.addAll(['-s', sourcePath, patchTmp, outputPath]);

        try {
          final res = Process.runSync(args.removeAt(0), args);
          if (res.exitCode == 0) {
            File(outputPath).renameSync(sourcePath);
            applied++;
            _log('[DCP xdelta] Patched: $targetName');
          } else {
            final err = res.stderr.toString().trim();
            _log('[DCP xdelta ERROR] $targetName: $err');
          }
        } catch (_) {
          _log('[DCP ERROR] xdelta3 binary execution failed.');
          break;
        }
      }
    } finally {
      try {
        tempDir.deleteSync(recursive: true);
      } catch (_) {}
    }

    return applied;
  }

  void _copyDisc(String srcDir, String dstDir) {
    final dir = Directory(srcDir);
    for (final entity in dir.listSync(recursive: true)) {
      if (entity is File) {
        final rel = p.relative(entity.path, from: srcDir);
        final dstPath = p.join(dstDir, rel);
        File(dstPath)
          ..createSync(recursive: true)
          ..writeAsBytesSync(entity.readAsBytesSync());
      }
    }
  }
}
